### PR TITLE
[wip] Add SAML support

### PR DESCRIPTION
**Pull request by ddurieux**

### CODE REVIEW

**Comment by remicollet** [comment]
Il y a clairement des trucs qui n'ons pas besoin d'être dans "lib" (doc, tests...)

A noter: il peut être sympa d'ajouter "onelogin/php-saml" dans tools/composer.json et d'adapter la provédure de mise à jour (update_lib.sh).

Notice: cette lib introduit une dépendance sur "mcrypt" ce qui me semble une très très mauvaise idée.

**Comment by ddurieux** [comment]
ok, merci pour les infos ;)
Quel est le problème avec mcrypt ?

**Comment by remicollet** [comment]
Read: http://blog.remirepo.net/post/2015/07/07/About-libmcrypt-and-php-mcrypt
Lire: http://blog.remirepo.net/post/2015/07/07/A-propos-de-libmcrypt-et-php-mcrypt

File: inc/auth.class.php [hunk 1]
```diff
@@ -411,6 +415,32 @@ function getAlternateAuthSystemsUserLogin($authtype=0) {
             $this->user->fields['name'] = phpCAS::getUser();
             return true;
 
+         case self::SAML :
+            require_once(GLPI_ROOT.'/lib/php-saml/_toolkit_loader.php');
```
Comment (remicollet) [inline inc/auth.class.php]: Merci de prévoir une redirection pour distribution downstream

            require_once(GLPI_PHP_SAML_AUTOLOADER); 

à définir dans config/based_config.php

File: lib/php-saml/CHANGELOG [hunk 2]
```diff
@@ -0,0 +1,59 @@
+CHANGELOG
```
Comment (remicollet) [inline lib/php-saml/CHANGELOG]: Why version 2.4.0 which is 1 year old ?
Upstream have released 2.50, 2.6.0, 2.6.1, 2.7.0.

File: inc/auth.class.php [hunk 3]
```diff
@@ -411,6 +415,32 @@ function getAlternateAuthSystemsUserLogin($authtype=0) {
...
+            $rep = base64_decode($_POST['SAMLResponse']);
+            $document = new DOMDocument();
+            $document = OneLogin_Saml2_Utils::loadXML($document, $rep);
+            $auth->processResponse();
+            $errors = $auth->getErrors();
+            if (!empty($auth->_errors)) {
+               print_r($auth->_errors);
+            }
+            if (!empty($errors)) {
+               print_r('SAML error: <p>'.implode(', ', $errors).'</p>');
```
Comment (remicollet) [inline inc/auth.class.php]: Should probably goes to a log file.

File: inc/authsaml.class.php [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+   function generate_settings() {
+      $this->getFromDB(1);
+
+      $sp_x509cert = $this->fields['sp_x509cert'];
+      $sp_x509cert = str_replace('-----BEGIN CERTIFICATE-----', '', $sp_x509cert);
+      $sp_x509cert = str_replace('-----END CERTIFICATE-----', '', $sp_x509cert);
+      $sp_x509cert = str_replace(array("\r\n", "\r", "\n"), "", $sp_x509cert);
+      $sp_x509cert = trim($sp_x509cert);
+
+$privkey = "-----BEGIN RSA PRIVATE KEY-----
```
Comment (remicollet) [inline inc/authsaml.class.php]: Where is key from ?
Should probably be store in a more secure way (config dir) and generated on install.

**Comment by wawax** [comment]
any news from this pr ?

**Comment by ddurieux** [comment]
I have fixed many bugs and made modification with @remicollet best practices ;)

Will try to push that next week

**Comment by wawax** [comment]
Hi David,

This is a definitly as must have feature, and we would be happy to see it coming into GLPi.
Do you plan to work on it soon ?

**Comment by ddurieux** [comment]
Not know, will see with @orthagh next week with 9.2 features

**Comment by orthagh** [comment]
@ddurieux Do you plan to continue this PR or i close it ?

**Comment by falcon78921** [comment]
I'm just curious..

Is integrating SAML into GLPI still an active project? 

I use GLPI 9.1.1 and have tried using the glpi-fpsaml plugin. However, the plugin doesn't seem to work. The plugin relies on outdated dependencies. 

Again, I would love to see this in GLPI someday! :)

**Comment by orthagh** [comment]
As we no longer have anyone working on this PR/subject, i archive it.

@falcon78921 i juste could redirect you to the suggestion topic: http://glpi.userecho.com/topics/202-saml-support-for-authentication/

**Pull request closed without merging**